crates/subspace-farmer/src/utils/archival_storage_info.rs

```diff
@@ -41,34 +42,30 @@ impl ArchivalStorageInfo {
         peer_filters.insert(peer_id, cuckoo_filter);
 
         // Truncate current peer set by limits.
-        let mut connected_peers = peer_filters.keys().cloned().collect::<Vec<_>>();
-
-        let exceeding_number_of_connected_peers = connected_peers
-            .len()
-            .saturating_div(CONNECTED_PEERS_NUMBER_LIMIT);
-
         let mut rng = StdRng::seed_from_u64({
             // Hash of PeerID
             let mut s = DefaultHasher::new();
             peer_id.hash(&mut s);
             s.finish()
         });
 
-        // Remove random peers when we exceed the limit of storing peers (and their cuckoo-filters).
-        for _ in 0..exceeding_number_of_connected_peers {
+        // Remove random peer when we exceed the limit of storing peers (and their cuckoo-filters).
+        if peer_filters.len() > CONNECTED_PEERS_NUMBER_LIMIT {
+            let connected_peers = peer_filters.keys().cloned().collect::<Vec<_>>();
             let random_index = rng.gen_range(0..connected_peers.len());
 
             let removing_peer_id = *connected_peers
                 .get(random_index)
                 .expect("Index is checked to be present.");
 
-            connected_peers.swap_remove(random_index);
             peer_filters.remove(&removing_peer_id);
+
+            debug!(%removing_peer_id, "Removed disconnected peer from filter cache.");
         }
```

Comment:
If we remove this limit we'll be dependent on the implicit limit of the `connected peers` protocol. IMHO It's better to manage limits on connections and memory separately.

Reply:
Why is it better? It makes no sense to me to have limits of a similar nature in multiple places. This thing should only store peer info of connected peers. It doesn't control who connects or how many of them, it is just a storage. I feel you're mixing responsibilities here unnecessarily and adding extra logic which already exists elsewhere.

Reply:
If we want to change connected-peers limit we could forget about this implicit dependency. Moreover, we could have different logic imposed on the filter cache limits (don't remove disconnected peers, have fewer filters than connected peers, etc). Overall, a separate limit seems to enable more agility.

Reply:
> If we want to change connected-peers limit we could forget about this implicit dependency

And? If we don't have a constant hardcoded here it doesn't matter, nothing needs to be changed here. And it is absolutely a reasonable assumption that if you have more connections you'll have proportionally higher memory usage. By introducing constants and additional logic here you make it way more complex because multiple constants in different places now likely need to be updated in sync with each other.

> Moreover, we could have different logic imposed on the filter cache limits (don't remove disconnected peers, have fewer filters than connected peers, etc).

Sounds interesting, but right now I don't see why we want that.

> Overall, a separate limit seems to enable more agility.

Well, I call it unnecessary complexity. We have total limits on libp2p level (we still do, right?), then connected peers limits and now this. It is way too much. Less limits, more things that "just work and do the right thing". The more moving parts you introduce the more likely they'll happen to be misaligned.